mysql2 does not build with mysql-client 8.3.0

Compiling mysql2 with `mysql-client` 8.3.0 fails:

```bash
❯ make
compiling client.c
In file included from client.c:15:
./mysql_enc_name_to_ruby.h:43:1: warning: a function definition without a prototype is deprecated in all versions of C and is not supported in C2x [-Wdeprecated-non-prototype]
mysql2_mysql_enc_name_to_rb_hash (str, len)
^
./mysql_enc_name_to_ruby.h:86:1: warning: a function definition without a prototype is deprecated in all versions of C and is not supported in C2x [-Wdeprecated-non-prototype]
mysql2_mysql_enc_name_to_rb (str, len)
^
client.c:1438:3: error: call to undeclared function 'mysql_ssl_set'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
  mysql_ssl_set(wrapper->client,
  ^
client.c:1438:3: note: did you mean 'mysql_close'?
/opt/homebrew/opt/mysql-client/include/mysql/mysql.h:797:14: note: 'mysql_close' declared here
void STDCALL mysql_close(MYSQL *sock);
             ^
2 warnings and 1 error generated.
make: *** [client.o] Error 1
```

`mysql_ssl_set` was [deprecated on 8.2 and removed on 8.3](https://dev.mysql.com/doc/relnotes/mysql/8.3/en/news-8-3-0.html).

@xjunior 

I follow your idea to install mysql-client 8.2. But it did not help me out:
```
Installing mysql2 0.5.5 with native extensions
Gem::Ext::BuildError: ERROR: Failed to build gem native extension.

    current directory: /Users/xinlangzi/.rvm/gems/ruby-3.1.3/gems/mysql2-0.5.5/ext/mysql2
/Users/xinlangzi/.rvm/rubies/ruby-3.1.3/bin/ruby -I /Users/xinlangzi/.rvm/rubies/ruby-3.1.3/lib/ruby/3.1.0 extconf.rb
--with-opt-dir\=/opt/homebrew/opt/openssl@3
checking for rb_absint_size()... yes
checking for rb_absint_singlebit_p()... yes
checking for rb_gc_mark_movable()... yes
checking for rb_wait_for_single_fd()... yes
checking for rb_enc_interned_str() in ruby.h... yes
-----
Using --with-openssl-dir=/opt/homebrew/opt/openssl@3
-----
-----
Using mysql_config at /opt/homebrew/bin/mysql_config
-----
checking for mysql.h... yes
checking for errmsg.h... yes
checking for SSL_MODE_DISABLED in mysql.h... yes
checking for SSL_MODE_PREFERRED in mysql.h... yes
checking for SSL_MODE_REQUIRED in mysql.h... yes
checking for SSL_MODE_VERIFY_CA in mysql.h... yes
checking for SSL_MODE_VERIFY_IDENTITY in mysql.h... yes
checking for MYSQL.net.vio in mysql.h... yes
checking for MYSQL.net.pvio in mysql.h... no
checking for MYSQL_DEFAULT_AUTH in mysql.h... yes
checking for MYSQL_ENABLE_CLEARTEXT_PLUGIN in mysql.h... yes
checking for SERVER_QUERY_NO_GOOD_INDEX_USED in mysql.h... yes
checking for SERVER_QUERY_NO_INDEX_USED in mysql.h... yes
checking for SERVER_QUERY_WAS_SLOW in mysql.h... yes
checking for MYSQL_OPTION_MULTI_STATEMENTS_ON in mysql.h... yes
checking for MYSQL_OPTION_MULTI_STATEMENTS_OFF in mysql.h... yes
checking for my_bool in mysql.h... no
-----
Don't know how to set rpath on your system, if MySQL libraries are not in path mysql2 may not load
-----
-----
Setting libpath to /opt/homebrew/Cellar/mysql/8.3.0/lib
-----
creating Makefile

current directory: /Users/xinlangzi/.rvm/gems/ruby-3.1.3/gems/mysql2-0.5.5/ext/mysql2
make DESTDIR\= sitearchdir\=./.gem.20240131-58086-wij5w2 sitelibdir\=./.gem.20240131-58086-wij5w2 clean

current directory: /Users/xinlangzi/.rvm/gems/ruby-3.1.3/gems/mysql2-0.5.5/ext/mysql2
make DESTDIR\= sitearchdir\=./.gem.20240131-58086-wij5w2 sitelibdir\=./.gem.20240131-58086-wij5w2
compiling client.c
In file included from client.c:15:
./mysql_enc_name_to_ruby.h:43:1: warning: a function definition without a prototype is deprecated in all versions of C and is not supported in C2x
[-Wdeprecated-non-prototype]
mysql2_mysql_enc_name_to_rb_hash (str, len)
^
./mysql_enc_name_to_ruby.h:86:1: warning: a function definition without a prototype is deprecated in all versions of C and is not supported in C2x
[-Wdeprecated-non-prototype]
mysql2_mysql_enc_name_to_rb (str, len)
^
client.c:1438:3: error: call to undeclared function 'mysql_ssl_set'; ISO C99 and later do not support implicit function declarations
[-Wimplicit-function-declaration]
  mysql_ssl_set(wrapper->client,
  ^
client.c:1438:3: note: did you mean 'mysql_close'?
/opt/homebrew/Cellar/mysql/8.3.0/include/mysql/mysql.h:797:14: note: 'mysql_close' declared here
void STDCALL mysql_close(MYSQL *sock);
             ^
2 warnings and 1 error generated.
make: *** [client.o] Error 1

make failed, exit code 2

Gem files will remain installed in /Users/xinlangzi/.rvm/gems/ruby-3.1.3/gems/mysql2-0.5.5 for inspection.
Results logged to /Users/xinlangzi/.rvm/gems/ruby-3.1.3/extensions/arm64-darwin-23/3.1.0/mysql2-0.5.5/gem_make.out

  /Users/xinlangzi/.rvm/rubies/ruby-3.1.3/lib/ruby/3.1.0/rubygems/ext/builder.rb:102:in `run'
  /Users/xinlangzi/.rvm/rubies/ruby-3.1.3/lib/ruby/3.1.0/rubygems/ext/builder.rb:51:in `block in make'
  /Users/xinlangzi/.rvm/rubies/ruby-3.1.3/lib/ruby/3.1.0/rubygems/ext/builder.rb:43:in `each'
  /Users/xinlangzi/.rvm/rubies/ruby-3.1.3/lib/ruby/3.1.0/rubygems/ext/builder.rb:43:in `make'
  /Users/xinlangzi/.rvm/rubies/ruby-3.1.3/lib/ruby/3.1.0/rubygems/ext/ext_conf_builder.rb:42:in `build'
  /Users/xinlangzi/.rvm/rubies/ruby-3.1.3/lib/ruby/3.1.0/rubygems/ext/builder.rb:171:in `build_extension'
  /Users/xinlangzi/.rvm/rubies/ruby-3.1.3/lib/ruby/3.1.0/rubygems/ext/builder.rb:205:in `block in build_extensions'
  /Users/xinlangzi/.rvm/rubies/ruby-3.1.3/lib/ruby/3.1.0/rubygems/ext/builder.rb:202:in `each'
  /Users/xinlangzi/.rvm/rubies/ruby-3.1.3/lib/ruby/3.1.0/rubygems/ext/builder.rb:202:in `build_extensions'
  /Users/xinlangzi/.rvm/rubies/ruby-3.1.3/lib/ruby/3.1.0/rubygems/installer.rb:843:in `build_extensions'
  /Users/xinlangzi/.rvm/gems/ruby-3.1.3/gems/bundler-2.4.20/lib/bundler/rubygems_gem_installer.rb:72:in `build_extensions'
  /Users/xinlangzi/.rvm/gems/ruby-3.1.3/gems/bundler-2.4.20/lib/bundler/rubygems_gem_installer.rb:28:in `install'
  /Users/xinlangzi/.rvm/gems/ruby-3.1.3/gems/bundler-2.4.20/lib/bundler/source/rubygems.rb:203:in `install'
  /Users/xinlangzi/.rvm/gems/ruby-3.1.3/gems/bundler-2.4.20/lib/bundler/installer/gem_installer.rb:54:in `install'
  /Users/xinlangzi/.rvm/gems/ruby-3.1.3/gems/bundler-2.4.20/lib/bundler/installer/gem_installer.rb:16:in `install_from_spec'
  /Users/xinlangzi/.rvm/gems/ruby-3.1.3/gems/bundler-2.4.20/lib/bundler/installer/parallel_installer.rb:156:in `do_install'
  /Users/xinlangzi/.rvm/gems/ruby-3.1.3/gems/bundler-2.4.20/lib/bundler/installer/parallel_installer.rb:147:in `block in worker_pool'
  /Users/xinlangzi/.rvm/gems/ruby-3.1.3/gems/bundler-2.4.20/lib/bundler/worker.rb:62:in `apply_func'
  /Users/xinlangzi/.rvm/gems/ruby-3.1.3/gems/bundler-2.4.20/lib/bundler/worker.rb:57:in `block in process_queue'
  /Users/xinlangzi/.rvm/gems/ruby-3.1.3/gems/bundler-2.4.20/lib/bundler/worker.rb:54:in `loop'
  /Users/xinlangzi/.rvm/gems/ruby-3.1.3/gems/bundler-2.4.20/lib/bundler/worker.rb:54:in `process_queue'
  /Users/xinlangzi/.rvm/gems/ruby-3.1.3/gems/bundler-2.4.20/lib/bundler/worker.rb:90:in `block (2 levels) in create_threads'

An error occurred while installing mysql2 (0.5.5), and Bundler cannot continue.
```

Did I miss other steps?

--- a/ext/mysql2/client.c
+++ b/ext/mysql2/client.c
@@ -1435,12 +1435,31 @@ static VALUE set_charset_name(VALUE self, VALUE value) {
 static VALUE set_ssl_options(VALUE self, VALUE key, VALUE cert, VALUE ca, VALUE capath, VALUE cipher) {
   GET_CLIENT(self);
 
+#ifdef HAVE_MYSQL_SSL_SET
   mysql_ssl_set(wrapper->client,
       NIL_P(key)    ? NULL : StringValueCStr(key),
       NIL_P(cert)   ? NULL : StringValueCStr(cert),
       NIL_P(ca)     ? NULL : StringValueCStr(ca),
       NIL_P(capath) ? NULL : StringValueCStr(capath),
       NIL_P(cipher) ? NULL : StringValueCStr(cipher));
+#else
+  /* mysql 8.3 does not provide mysql_ssl_set */
+  if (NIL_P(key)) {
+    mysql_options(wrapper->client, MYSQL_OPT_SSL_KEY, StringValueCStr(key));
+  }
+  if (NIL_P(cert)) {
+    mysql_options(wrapper->client, MYSQL_OPT_SSL_CERT, StringValueCStr(cert));
+  }
+  if (NIL_P(ca)) {
+    mysql_options(wrapper->client, MYSQL_OPT_SSL_CA, StringValueCStr(ca));
+  }
+  if (NIL_P(capath)) {
+    mysql_options(wrapper->client, MYSQL_OPT_SSL_CAPATH, StringValueCStr(capath));
+  }
+  if (NIL_P(cipher)) {
+    mysql_options(wrapper->client, MYSQL_OPT_SSL_CIPHER, StringValueCStr(cipher));
+  }
+#endif
 
   return self;
 }

--- a/ext/mysql2/extconf.rb
+++ b/ext/mysql2/extconf.rb
@@ -164,6 +164,9 @@ have_const('MYSQL_OPTION_MULTI_STATEMENTS_OFF', mysql_h)
 # to retain compatibility with the typedef in earlier MySQLs.
 have_type('my_bool', mysql_h)
 
+# detect mysql functions
+have_func('mysql_ssl_set', mysql_h)
+
 ### Compiler flags to help catch errors
 
 # This is our wishlist. We use whichever flags work on the host.

